File: 2017/python/day2.py
import numpy as np

with open('../input/input-day2.txt', 'r') as input_file:
    part1 = 0
    part2 = 0
    for line in input_file.readlines():
        numbers = [int(x) for x in line.split('\t')]
        part1 += np.max(numbers) - np.min(numbers)

        for x in numbers:
            for y in numbers:
                if x == y: continue
                if x % y == 0:
                    part2 += int(x / y)

print("Part 1: " + str(part1))
print("Part 2: " + str(part2))

File: 2020/typescript/src/days/day2.ts
import {Day} from "../aoc/day";

export class Day2 extends Day {
    day = (): number => 2;

    part1 = (): string => this.countWith((passwd, char, min, max) =>
        Day2.charCount(char, passwd) >= min && Day2.charCount(char, passwd) <= max
    );

    part2 = (): string => this.countWith((pass, c, a, b) =>
        (pass[a - 1] === c || pass[b - 1] === c) && !(pass[a - 1] === c && pass[b - 1] === c));

    private countWith(evaluate: { (passwd: string, char: string, min: number, max: number): boolean }): string {
        return this.input
            .map((l) => l.match(/^(?<min>\d+)-(?<max>\d+) (?<char>\w): (?<passwd>\w+)$/))
            .filter(match => (match === null || match.groups === undefined)
                ? false
                : evaluate(
                    match.groups['passwd'],
                    match.groups['char'],
                    parseInt(match.groups['min']),
                    parseInt(match.groups['max'])
                )
            )
            .length
            .toString();
    }

    private static charCount(char: string, subject: string): number {
        let result = 0;
        for (let c of subject) {
            result += Number(c === char);
        }

        return result
    }
}